modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotFutureTask.java

```diff
@@ -0,0 +1,840 @@
+        return Objects.hash(snpName);
+    }
+
+    /** {@inheritDoc} */
+    @Override public String toString() {
+        return S.toString(SnapshotFutureTask.class, this);
+    }
+
+    /**
+     *
+     */
+    private static class PageStoreSerialWriter implements PageWriteListener, Closeable {
+        /** Ignite logger to use. */
+        @GridToStringExclude
+        private final IgniteLogger log;
+
+        /** Page store to which current writer is related to. */
+        private final PageStore store;
+
+        /** Busy lock to protect write opertions. */
+        private final ReadWriteLock lock = new ReentrantReadWriteLock();
+
+        /** Local buffer to perpform copy-on-write operations. */
+        private final ThreadLocal<ByteBuffer> localBuff;
+
+        /** {@code true} if need the original page from PageStore instead of given buffer. */
+        private final BooleanSupplier checkpointComplete;
+
+        /** {@code true} if snapshot process is stopping or alredy stopped. */
+        private final BooleanSupplier interrupt;
+
+        /** Callback to stop snapshot if an error occurred. */
+        private final Consumer<Throwable> exConsumer;
+
+        /** IO over the underlying file */
+        private volatile FileIO fileIo;
+
+        /** {@code true} if partition file has been copied to external resource. */
+        private volatile boolean partProcessed;
+
+        /** {@code true} means current writer is allowed to handle page writes. */
+        private volatile boolean inited;
+
+        /**
+         * Array of bits. 1 - means pages written, 0 - the otherwise.
+         * Size of array can be estimated only under checkpoint write lock.
+         */
+        private volatile AtomicBitSet pagesWrittenBits;
+
+        /**
+         * @param log Ignite logger to use.
+         * @param checkpointComplete Checkpoint finish flag.
+         * @param pageSize Size of page to use for local buffer.
+         * @param cfgFile Configuration file provider.
+         * @param factory Factory to produce an IO interface over underlying file.
+         */
+        public PageStoreSerialWriter(
+            IgniteLogger log,
+            PageStore store,
+            BooleanSupplier checkpointComplete,
+            BooleanSupplier interrupt,
+            Consumer<Throwable> exConsumer,
+            File cfgFile,
+            FileIOFactory factory,
+            int pageSize
+        ) throws IgniteCheckedException {
+            assert store != null;
+
+            try {
+                fileIo = factory.create(cfgFile);
+            }
+            catch (IOException e) {
+                throw new IgniteCheckedException(e);
+            }
+
+            this.checkpointComplete = checkpointComplete;
+            this.interrupt = interrupt;
+            this.exConsumer = exConsumer;
+            this.log = log.getLogger(PageStoreSerialWriter.class);
+
+            localBuff = ThreadLocal.withInitial(() ->
```

Comment:
You need only 1 buffer per thread, but you create buffer per page store per thread, it's redundant and can lead to OOM errors (direct buffer deallocates only when ByteBuffer is garbage collected, but you can get out of off-heap memory before it).